File: bindings/penta-ffi/smoke.c
/* Smoke test for the C ABI: plays full games against both built-in bots
 * choosing pseudo-random legal actions, and checks the JSON surface looks
 * like the protocol. Run via scripts/check-bindings.sh. */

#include "include/penta.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static unsigned long rng_state = 12345;

static unsigned long next_rand(void) {
    /* Deterministic LCG so the smoke test never flakes. */
    rng_state = rng_state * 6364136223846793005UL + 1442695040888963407UL;
    return rng_state >> 33;
}

static int fail(const char *what) {
    fprintf(stderr, "FAIL: %s: %s\n", what, penta_last_error());
    return 1;
}

/* Plays one uniform-random action on game, and the same index on mirror
 * when mirror is not NULL, so two copies can be driven in lockstep. */
static int act_random(PentaGame *game, PentaGame *mirror) {
    uint32_t count = penta_legal_action_count(game);
    if (count == 0) {
        fprintf(stderr, "FAIL: no legal actions but no result\n");
        return 1;
    }
    uint32_t pick = (uint32_t)(next_rand() % count);
    if (penta_act(game, pick) != 0)
        return fail("penta_act");
    if (mirror && penta_act(mirror, pick) != 0)
        return fail("penta_act on a clone");
    return 0;
}

static int play_one(const char *config, int check_json) {
    PentaGame *game = penta_new(config);
    if (!game) return fail("penta_new");

    if (check_json) {
        int32_t seat = penta_decision_seat(game);
        char *observation = penta_observe_json(game, seat);
        if (!observation) return fail("penta_observe_json");
        if (!strstr(observation, "\"legalActions\"") ||
            !strstr(observation, "\"seat\"") ||
            !strstr(observation, "\"protocolVersion\"") ||
            !strstr(observation, "\"simulationFingerprint\"") ||
            !strstr(observation, penta_simulation_fingerprint())) {
            fprintf(stderr, "FAIL: observation missing protocol fields\n");
            return 1;
        }
        penta_string_free(observation);
    }

    /* Uniform random over the whole list. Nothing in it resigns, so this
     * plays a real if witless game rather than ending on turn one. */
    int steps;
    for (steps = 0; steps < 200000; steps++) {
        if (penta_result(game) != -1) break;
        if (act_random(game, NULL) != 0) return 1;
    }

    int32_t result = penta_result(game);
    penta_free(game);
    if (result == -1) {
        fprintf(stderr, "FAIL: game did not finish in %d steps\n", steps);
        return 1;
    }
    printf("ok: result=%d after %d of your decisions\n", result, steps);
    return 0;
}

static int check_standard_game(void) {
    PentaGame *game = penta_new(
        "{\"format\":\"isd-rtr-standard\","
        "\"p1Deck\":\"Briksza Naya Midrange\","
        "\"p2Deck\":\"Greer G/R Aggro\","
        "\"opponent\":\"external\",\"seed\":17}");
    if (!game) return fail("penta_new Standard");

    int32_t seat = penta_decision_seat(game);
    char *observation = penta_observe_json(game, seat);
    if (!observation) {
        penta_free(game);
        return fail("penta_observe_json Standard");
    }
    int valid = strstr(observation, "\"format\":\"isd-rtr-standard\"") != NULL;
    penta_string_free(observation);
    penta_free(game);
    if (!valid) {
        fprintf(stderr, "FAIL: Standard observation has the wrong format\n");
        return 1;
    }
    return 0;
}

/* A cloned game replays identically under the same actions, and acting on
 * one copy never disturbs the other. */
static int check_clone(const char *config) {
    PentaGame *game = penta_new(config);
    if (!game) return fail("penta_new");

    /* Reach a mid-game state. */
    for (int step = 0; step < 30; step++) {
        if (act_random(game, NULL) != 0) return 1;
        if (penta_result(game) != -1) {
            fprintf(stderr, "FAIL: game ended before the clone check\n");
            return 1;
        }
    }

    /* Identical bytes at the fork, and after the same actions. */
    PentaGame *replay = penta_clone(game);
    if (!replay) return fail("penta_clone");
    for (int step = 0; step < 10; step++) {
        int32_t seat = penta_decision_seat(game);
        char *original = penta_observe_json(game, seat);
        char *cloned = penta_observe_json(replay, seat);
        if (!original || !cloned) return fail("penta_observe_json");
        if (strcmp(original, cloned) != 0) {
            fprintf(stderr, "FAIL: clone diverged under identical actions\n");
            return 1;
        }
        penta_string_free(original);
        penta_string_free(cloned);
        if (act_random(game, replay) != 0) return 1;
        if (penta_result(game) != penta_result(replay)) {
            fprintf(stderr, "FAIL: clone and original disagree on result\n");
            return 1;
        }
        if (penta_result(game) != -1) break;
    }
    penta_free(replay);

    /* Diverge: the fork plays a different legal action than the original,
     * the two games stop matching, and the original never notices. */
    if (penta_result(game) == -1) {
        /* Walk to a decision with at least two options to disagree on. */
        while (penta_result(game) == -1 && penta_legal_action_count(game) < 2) {
            if (act_random(game, NULL) != 0) return 1;
        }
        if (penta_result(game) != -1) {
            fprintf(stderr, "FAIL: game ended before the divergence check\n");
            return 1;
        }
        uint32_t count = penta_legal_action_count(game);
        uint32_t choice = (uint32_t)(next_rand() % count);
        uint32_t other = (choice + 1) % count;
        int32_t seat = penta_decision_seat(game);
        char *before = penta_observe_json(game, seat);
        if (!before) return fail("penta_observe_json");
        PentaGame *fork = penta_clone(game);
        if (!fork) return fail("penta_clone");
        if (penta_act(fork, other) != 0)
            return fail("penta_act on a fork");
        char *after = penta_observe_json(game, seat);
        if (!after) return fail("penta_observe_json");
        if (strcmp(before, after) != 0) {
            fprintf(stderr, "FAIL: the original changed when its fork acted\n");
            return 1;
        }
        penta_string_free(before);
        penta_string_free(after);
        if (penta_act(game, choice) != 0) return fail("penta_act");
        char *original = penta_observe_json(game, seat);
        char *forked = penta_observe_json(fork, seat);
        if (!original || !forked) return fail("penta_observe_json");
        if (strcmp(original, forked) == 0) {
            fprintf(stderr, "FAIL: different actions, same observation\n");
            return 1;
        }
        penta_string_free(original);
        penta_string_free(forked);
        /* A fork is a live game, not a snapshot: it plays on by itself. */
        for (int step = 0; step < 10; step++) {
            if (penta_result(fork) != -1) break;
            if (act_random(fork, NULL) != 0) return 1;
        }
        penta_free(fork);
    }
    penta_free(game);
    printf("clone: forks replay identically and diverge independently\n");
    return 0;
}

int main(void) {
    const char *fingerprint = penta_simulation_fingerprint();
    if (!fingerprint || strncmp(fingerprint, "sha256-", 7) != 0 ||
        strlen(fingerprint) != 71)
        return fail("penta_simulation_fingerprint");
    printf("engine %s, protocol %u, simulation %s\n", penta_engine_version(),
           penta_protocol_version(), fingerprint);

    char *decks = penta_deck_names_json();
    if (!decks || !strstr(decks, "Sligh")) return fail("penta_deck_names_json");
    penta_string_free(decks);

    char *standard_decks =
        penta_deck_names_for_format_json("isd-rtr-standard");
    if (!standard_decks || !strstr(standard_decks, "Briksza Naya Midrange"))
        return fail("penta_deck_names_for_format_json");
    penta_string_free(standard_decks);

    char *catalog = penta_catalog_json();
    if (!catalog || !strstr(catalog, "Lightning Bolt") ||
        !strstr(catalog, fingerprint))
        return fail("penta_catalog_json");
    penta_string_free(catalog);

    char *standard_catalog =
        penta_catalog_json_for_format("isd-rtr-standard");
    if (!standard_catalog ||
        !strstr(standard_catalog, "\"format\":\"isd-rtr-standard\"") ||
        !strstr(standard_catalog, "Huntmaster of the Fells"))
        return fail("penta_catalog_json_for_format");
    penta_string_free(standard_catalog);

    if (check_standard_game()) return 1;

    /* Random moves against each built-in opponent, and a self-play game. */
    if (play_one("{\"p1Deck\":\"Sligh\",\"p2Deck\":\"The Deck\","
                 "\"opponent\":\"handcrafted\",\"opponentSeat\":\"p2\","
                 "\"seed\":7}", 1))
        return 1;
    if (play_one("{\"p1Deck\":\"Goblins\",\"p2Deck\":\"White Weenie\","
                 "\"opponent\":\"random\",\"opponentSeat\":\"p2\","
                 "\"seed\":11}", 0))
        return 1;
    if (play_one("{\"p1Deck\":\"Sligh\",\"p2Deck\":\"Goblins\","
                 "\"opponent\":\"external\",\"seed\":13}", 0))
        return 1;

    /* Clones: fork a game mid-state and check the copies are independent. */
    if (check_clone("{\"p1Deck\":\"Sligh\",\"p2Deck\":\"The Deck\","
                    "\"opponent\":\"handcrafted\",\"opponentSeat\":\"p2\","
                    "\"seed\":7}"))
        return 1;

    /* Error paths report through penta_last_error instead of crashing. */
    if (penta_new("{\"p1Deck\":\"Not A Deck\",\"p2Deck\":\"Sligh\"}") != NULL) {
        fprintf(stderr, "FAIL: bad deck accepted\n");
        return 1;
    }
    if (strlen(penta_last_error()) == 0) {
        fprintf(stderr, "FAIL: bad deck left no error message\n");
        return 1;
    }
    if (penta_catalog_json_for_format("not-a-format") != NULL) {
        fprintf(stderr, "FAIL: bad format accepted\n");
        return 1;
    }
    if (strlen(penta_last_error()) == 0) {
        fprintf(stderr, "FAIL: bad format left no error message\n");
        return 1;
    }

    printf("smoke test passed\n");
    return 0;
}

File: bindings/penta-py/smoke.py
"""Smoke test for the Python bindings: full games, determinism, self-play.

Run via scripts/check-bindings.sh, which builds the module and puts it on
the path first.
"""

import json

import penta

fingerprint = penta.simulation_fingerprint()
assert fingerprint.startswith("sha256-")
assert len(fingerprint) == 71
assert all(character in "0123456789abcdef" for character in fingerprint[7:])
print(
    "engine",
    penta.engine_version(),
    "protocol",
    penta.protocol_version(),
    "simulation",
    fingerprint,
)
assert "Sligh" in penta.deck_names()
catalog_payload = json.loads(penta.catalog())
assert catalog_payload["simulationFingerprint"] == fingerprint
catalog = {c["definition"]: c for c in catalog_payload["cards"]}
assert any(c["name"] == "Lightning Bolt" for c in catalog.values())

standard_decks = penta.deck_names(format="isd-m14-standard")
assert "Briksza Naya Midrange" in standard_decks
standard_catalog_payload = json.loads(penta.catalog(format="isd-m14-standard"))
assert standard_catalog_payload["format"] == "isd-m14-standard"
assert standard_catalog_payload["simulationFingerprint"] == fingerprint
assert any(
    card["name"] == "Huntmaster of the Fells // Ravager of the Fells"
    for card in standard_catalog_payload["cards"]
)

standard_game = penta.Game(
    "Briksza Naya Midrange",
    "Greer G/R Aggro",
    opponent="external",
    format="isd-m14-standard",
    seed=17,
)
standard_observation = json.loads(standard_game.observe())
assert standard_observation["format"] == "isd-m14-standard"
assert standard_observation["simulationFingerprint"] == fingerprint

try:
    penta.deck_names(format="not-a-format")
except ValueError:
    pass
else:
    raise AssertionError("bad format accepted")

def pass_bot(obs):
    prefer = ["KeepHand", "ChooseDecision", "PassPriority", "FinishDeclaringAttackers",
              "FinishDeclaringBlockers", "AssignCombatDamage", "DiscardCards",
              "BottomCards", "ChooseUntap"]
    actions = obs["legalActions"]
    for kind in prefer:
        for action in actions:
            if action["type"] == kind:
                return action["index"]
    return 0

# vs handcrafted
game = penta.Game("Sligh", "The Deck", opponent="handcrafted", seed=7)
steps = 0
while game.result() is None:
    obs = json.loads(game.observe())
    game.act(pass_bot(obs))
    steps += 1
    assert steps < 100000
print("vs handcrafted:", game.result(), "in", steps, "decisions")

# determinism
def run(seed):
    g = penta.Game("Goblins", "Sligh", opponent="random", seed=seed)
    trace = []
    while g.result() is None:
        obs = json.loads(g.observe())
        trace.append(len(obs["legalActions"]))
        g.act(pass_bot(obs))
    return g.result(), trace
a, b = run(99), run(99)
assert a == b, "same seed, same game"
print("determinism ok:", a[0], "over", len(a[1]), "decisions")

# clone: a fork replays identically and diverges independently
game = penta.Game("Sligh", "The Deck", opponent="handcrafted", seed=7)
for _ in range(30):
    game.act(pass_bot(json.loads(game.observe())))
replay = game.clone()
assert game.observe() == replay.observe(), "a clone starts byte-identical"
for _ in range(20):
    choice = pass_bot(json.loads(game.observe()))
    game.act(choice)
    replay.act(choice)
    assert game.observe("p1") == replay.observe("p1"), "same actions, same bytes"
# Diverge: the fork plays a different legal action than the original, the
# two games stop matching, and the original never notices. Walk to a
# decision with at least two options first.
while len(json.loads(game.observe())["legalActions"]) < 2:
    game.act(0)
obs = json.loads(game.observe())
choice = pass_bot(obs)
other = (choice + 1) % len(obs["legalActions"])
before = game.observe()
fork = game.clone()
fork.act(other)
assert game.observe() == before, "the original is untouched"
game.act(choice)
assert game.observe("p1") != fork.observe("p1"), \
    "different actions, different games"
for _ in range(10):  # a fork is a live game, not a snapshot: it plays on
    if fork.result() is not None:
        break
    fork.act(pass_bot(json.loads(fork.observe())))
print("clone: forks replay identically and diverge independently")

# self-play: one loop drives both seats
game = penta.Game("Goblins", "White Weenie", opponent="external", seed=13)
steps = 0
while game.result() is None:
    seat = game.decision_seat()
    obs = json.loads(game.observe(seat))
    assert obs["seat"] == seat
    game.act(pass_bot(obs))
    steps += 1
    assert steps < 200000
print("self-play:", game.result(), "in", steps, "decisions")

# A hosted observation can become a live local determinization without the
# host seed. Here the hypotheses happen to use a local game's true hidden
# hand and libraries only so the smoke test can obtain correctly sized lists
# cheaply. Outside-game lists are explicit hypotheses and may be empty.
game = penta.Game("Sligh", "The Deck", opponent="external", seed=31)
view_json = game.observe("p1")
view = json.loads(view_json)
zone_definitions = lambda zone: [card["definition"] for card in json.loads(zone)]
hidden = {
    "hands": {"p2": zone_definitions(game.hand("p2"))},
    "libraries": {
        "p1": zone_definitions(game.library("p1")),
        "p2": zone_definitions(game.library("p2")),
    },
    "outsideGame": {"p1": [], "p2": []},
}
world = penta.Game.from_observation(view_json, json.dumps(hidden), rollout_seed=999)
rebuilt = json.loads(world.observe("p1"))
assert rebuilt["hand"] == view["hand"], "public object ids survive reconstruction"
assert rebuilt["legalActions"] == view["legalActions"]
assert "seed" not in rebuilt and "rng" not in rebuilt["checkpoint"]
world.act(0)
print("checkpoint: redacted observation rebuilt as a live local world")


# hypothetical worlds: you do not know what the opponent holds, so build the
# worlds you think are plausible and roll each one out. The engine supplies no
# distribution -- naming the cards is the whole API.
by_name = {card["name"]: definition for definition, card in catalog.items()}

game = penta.Game("Sligh", "The Deck", opponent="handcrafted", seed=3)
game.act(pass_bot(json.loads(game.observe())))
truth = json.loads(game.hand("p2"))
assert truth, "the simulation view shows the opponent's real hand"

worlds = []
for guess in ("Lightning Bolt", "Counterspell"):
    world = game.clone()
    world.set_hand("p2", [by_name["Mountain"], by_name[guess]])
    held = [c["definition"] for c in json.loads(world.hand("p2"))]
    assert held == [by_name["Mountain"], by_name[guess]]
    world.act(pass_bot(json.loads(world.observe())))
    worlds.append(held)

assert worlds[0] != worlds[1], "two different worlds from one position"
assert json.loads(game.hand("p2")) == truth, "the real game is untouched"

# A library can be stacked outright, including down to nothing.
world = game.clone()
world.set_library("p2", [by_name["Black Lotus"]])
assert len(json.loads(world.library("p2"))) == 1
world.set_library("p2", [])
assert json.loads(world.library("p2")) == []

try:
    world.set_hand("p2", [60000])
    raise AssertionError("a card outside the catalog must raise")
except ValueError:
    pass
print("simulate: opponent hands rewritten, worlds diverge, real game untouched")

# hidden info: p1 never sees p2's hand
game = penta.Game("Sligh", "The Deck", opponent="handcrafted", seed=3)
obs = json.loads(game.observe())
assert "opponentHandSize" in obs and isinstance(obs["opponentHandSize"], int)
print("smoke test passed")
